### ISSUE TITLE
Make ChatLog refer to messages by ChatLogIdx, dynamically remove messages from view, and request new messages from IChatLog

### ISSUE THREAD

**Issue by anthonybilinski** [qTox/qTox#6223]
##### Brief Description

This issue depends on https://github.com/qTox/qTox/issues/6222.

After #6222 is complete, every message that ChatLog displays will have a uniquely identifying ID, ChatLogIdx, which IChatLog will be aware of. This gives the ability for only ChatLog to care about what messages are currently in view, rather than GenericChatForm having to track it as well. ChatLog can then request messages to be loaded by just calling an API like "load messages from ChatLogIdx A to ChatLogIdx B" which IChatLog can generically fulfil.

This should be able to remove a lot of responsibility from GenericChatForm, and will heavily change the API of ChatLog. ChatLog will need to keep track of the ChatLogIdx of each message it has to be able to know which message indexes come after or before on scroll.

With this change, ChatLog's responsibility should be purely view, and should be the sole class responsibility of which messages are currently loaded. Because of that, ChatLog should be able to remove messages from view on scroll without needing to update any other classes.

With this overhaul of ChatLog, it could maybe be also be renamed to something like ChatView to reduce confusion with IChatLog which is a model concept but which has a very similar name.

**Comment by sphaerophoria** [comment]
I've been investigating this a bit in the past few days and just wanted to dump my thoughts in case others had input...

Looking at ChagLog/GenericChatForm it looks like the intended design here is that ChatLog is analogous to "views" in Qt. This class is meant to handle layout and display of chat messages.

Where it gets a little tricky is that GenericChatForm takes on a lot of the responsibilities that are ChatLog layout related. Not only is it informing ChatLog of the new messages, but it's converting them into ChatLinePtrs first. It's also handling the current scrolled view, and probably some other pieces that I've missed.

I was reading through some of the Model/View programming Qt docs [1] [2], and was experimenting with implementing a QAbstractItemModel and feeding it to a QTableView/QTreeView/QListView. I think in an ideal world this is how we would implement a ChatLog. It gives us a lot of what we want for free.

Unfortunately I do not think that the existing Qt widgets give enough flexibility for what we want, dynamic loading in both directions and text selection of underlying elements seems a little tricky with this approach. I think keeping the existing ChatLog with some extra tweaks may be a good way forward here. 

I'd like to have ChatLog behave in a way similar to QListView would in the following aspects

- ChatLog monitors an underlying model. Since we're writing this ourselves we can probably use IChatLog directly here instead of adapting it into something more like QAbstractItemModel
- ChatLog is in control of it's own viewports. With QListView/QAbstractItemModel you can implement a "fetchMore()" function on your model and when you hit the bottom of the view QListView will load it automatically. We can have ChatLog monitor it's own scroll position and handle loading IChatLog items as it needs to

Basically I'm agreeing with @anthonybilinski's original comment. ChatLog should be entirely a view class, and should pull from the underlying model so that GenericChatForm isn't involved in layout at all. GenericChatForm can still propagate insertions into the model if we want (analogous to the model insertion in [2]).

[1] https://doc.qt.io/qt-5/model-view-programming.html
[2] https://wiki.qt.io/WIP-How_to_create_a_simple_chat_application

**Comment by sphaerophoria** [comment]
I'm looping back on this, I think it probably is possible to get the behavior out of the Qt Model/View workflow. Something like https://stackoverflow.com/questions/33419892/can-i-select-text-by-mouse-from-a-number-of-delegates-in-qml sounds applicable. I know it's QML and not QtWidget related but I think the same concepts probably apply.

**Comment by sphaerophoria** [comment]
I've convinced myself yet again that flipping to a Model/View/Delegate model doesn't work for our needs. I keep thinking to myself "a chat log has to be one of the most basic use cases out there" but I don't think it is.

So just to summarize what I've tried so far and convince future me (and maybe others) that this is a bad idea, I'll just write through what I've come up with so far

* The general implementation would be to subclass QAbstractItemModel with something that converts our IChatLog messages into QVariants for display
* We would then need to implement a delegate to display our items
* This delegate gets a little frustrating...
    * We need to know the size of our item before we render it,this is so that the parent view can call sizeHint() and pre-allocate the space for us to render. 
        * This does not work well for us since we do not know the size of our chatline until we render it. Why? Emoji drawing, bold text, italics, etc. can all adjust the final size. Currently we use some basic HTML + QTextDocument to render this in advance.
        * If we're rendering in advance we might as well cache the rendered item for later
        * At that point we're basically rendering line by line and we've just re-invented the existing chatline pointer
    * Selection of items within a view is not supported out of the box. 
        * The best solution I could come up with was to use the "createEditor" concept to create an "editor" for the current selection. I used a QLabel that rendered the same text that I had already rendered in the paint function
        * This is theoretically cheap since the only time we need a custom widget is for the item that's been selected
        * We would need to save the selection to the model on release, and then re-paint the selection ourselves manually
        * We would need to enter edit mode on mouse click and pass the initial mouse click into the child widget, this seems possible, but I didn't see an easy way to extract the editor widget from the parent table to do so
        * Another option would be to use a QTextDocument or something to reverse lookup rendered text position to mouse position in the event handlers in the delegate. This would require us to store a rendered version of the document to do the reverse lookup, oh look we've re-invented ChatLine::Ptr again
   * Qts views only support dynamic loading in one direction, to support going in the other direction or purging rows we'd need to do some heavy index shuffling which would result in a full re-draw every time we load content in either direction
   * Basically it becomes clear that this Model/View/Delegate model is not designed for our usecase. For us we want to dynamically load the stuff at the end, and the stuff at the beginning which apparently requires a fully custom widget implementation. I'm not sure if using a QGraphicsScene is the right solution here, but after browsing the other options it doesn't seem as crazy as it initially looked.

Code of current experiments is currently at https://github.com/sphaerophoria/qTox/tree/6223-Model-View-Delegate

**Comment by sphaerophoria** [comment]
Alright, I think I'm fairly happy with. Just waiting on #6290, then the work from https://github.com/qTox/qTox/compare/master...sphaerophoria:6223-chatlog-cleanup can go in after I file a PR for #6222.

What I have should also resolve #5878 and #5940

**Comment by sphaerophoria** [comment]
There were a couple issues leftover with the previous implementation

1. Data consistency issues when selected chat lines were removed from the chatlog
2. Scrolling down was a little too jumpy due to double triggering of the scroll event

I've fixed both of these and after a loose review I think I'm happy with it again.
